src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -360,19 +357,30 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         match self.target_metadata(unit) {
             Some(ref metadata) => format!("{}{}", unit.target.crate_name(),
                                           metadata.extra_filename),
-            None if unit.target.allows_underscores() => {
-                unit.target.name().to_string()
-            }
-            None => unit.target.crate_name(),
+            None => self.bin_stem(unit),
+        }
+    }
+
+    fn bin_stem(&self, unit: &Unit) -> String {
+        let bin = if unit.target.allows_underscores() {
+            unit.target.name().to_string()
+        } else {
+            unit.target.crate_name()
+        };
+        if unit.profile.test {
+            bin + "-test"
```

Comment:
Hm I'm trying to recall why something like this would be required but I'm drawing a blank. Could you remind me why this is needed here?